### PR TITLE
Improve mem efficiency of constant folding

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108604
* __->__ #108421

Couple changes to make it more efficient.

- Because we replacing nodes that only have a single value, only store a single value instead of the whole tensor for node replacement
- torch.fx.Interpreter will preserve a Tensor in the env as long as it has more uses. That also applies even to output uses, but we are not going to constant fold that use. Instead of using last use for garbage collection, use last non output use. 

If reviewers would prefer I ghstack this bc of code movement let me know.

Fix for https://github.com/pytorch/pytorch/issues/108388

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D49020616](https://our.internmc.facebook.com/intern/diff/D49020616)

### CODE REVIEW

File: torch/_inductor/constant_folding.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                seen_uses.add(inp)
+                last_non_output_use[node].append(inp)
+
+            pytree.tree_map_only(torch.fx.Node, add_use, (node.args, node.kwargs))
+
+            # if this node is only used in output, we want to gc it right away
+            if len(node.users) == 1 and output_node in node.users:
+                last_non_output_use[node].append(node)
+
+        return last_non_output_use
```
Comment (eellison) [inline torch/_inductor/constant_folding.py:85]: This is only change in this file

**Comment by jansel** [comment]
Lints/test errors

It would be easier to review if the code movement were in another PR.

**Review by jansel — APPROVED** [review]

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49020616).

**Pull request closed without merging**